### PR TITLE
update trainer, base_model, add loss, optimizer

### PR DESCRIPTION
**Pull request by choosewhatulike**
- remove define_optimizer() and define_loss() from Trainer
Now Trainer get optimizer and loss_func from the input model

- add prepare_train() for BaseModel, which prepare required configures using in Trainer

- add name alias for loss & optimizer, we can switch to use fastNLP.core.optimizer instead of torch.optim

### CODE REVIEW

**Comment by FengZiYjun** [comment]
这样做还是挺巧妙的

**Comment by xpqiu** [comment]
建议optm和loss还是放在trainer里，和basemodel无关。trainer初始化时传给tainer。

**Comment by FengZiYjun** [comment]
@xpqiu @choosewhatulike 我做了一个这样的loss和optimizer，稍后push

**Comment by xpqiu** [comment]
close, use @FengZiYjun 's version.
